05_concurrency/homework/main.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type Cache interface {
+	Set(key string, value interface{}, ttl time.Duration)
+	Get(key string) (interface{}, bool)
+	Delete(key string)
+}
+
+type Object struct {
+	value     interface{}
+	timePoint time.Time     // Point of time when the user set value
+	ttl       time.Duration // Time To Live
+}
+
+type CacheStruct struct {
+	sync.Mutex
+	object map[string]Object
+}
+
+func (c *CacheStruct) Delete(key string) {
+	c.Lock()
+	defer c.Unlock()
+	delete(c.object, key)
+}
+
+func (c *CacheStruct) Get(key string) (interface{}, bool) {
+
+	c.Lock()
+	defer c.Unlock()
+
+	valueItem, ok := c.object[key]
+	if !ok {
+		return nil, false
+	}
+	// update the object
+	c.object[key] = Object{
+		value:     valueItem.value,
+		timePoint: time.Now(),
+		ttl:       c.object[key].ttl,
+	}
+	return valueItem.value, true
+}
+
+func (c *CacheStruct) Set(key string, value interface{}, ttl time.Duration) {
+	c.Lock()
+	defer c.Unlock()
+	c.object[key] = Object{
+		value:     value,
+		timePoint: time.Now(),
+		ttl:       ttl,
+	}
+	// create goroutine for every value
+	go func() {
+		for {
+			c.Lock()
```

Comment:
This lock will take too much CPU time and It significantly slowdown performance. Adding `<-time.After(time.Second*5)` will be good solution.